### PR TITLE
Improvements to SimpleHostRoutingFilter (continued)

### PR DESCRIPTION
**Pull request by liebowitz**
Hopefully this diff is more readable.  
Thanks.

### CODE REVIEW

**Comment by spencergibb** [comment]
@liebowitz have you signed the contributors agreement?

**Comment by spencergibb** [comment]
Explain to me how you fixed ignored headers?

**Comment by liebowitz** [comment]
Yes, agreement has been signed.

The RequestContext is stored in a thread local. When ignored headers are set in the config phase (e.g. below), they are not visible to the servlet executor thread pool.
```
@Configuration
@EnableZuulProxy
public class ZuulProxyConfig extends ZuulProxyConfiguration {
    @Bean
    @Override
    public SimpleHostRoutingFilter simpleHostRoutingFilter() {
        ProxyRequestHelper helper = new ProxyRequestHelper();
        helper.addIgnoredHeaders("X-Frame-Options");
        return new SimpleHostRoutingFilter(helper);
    }
}
```

**Comment by spencergibb** [comment]
I believe you've broken some existing functionality

	/**
	 * Zuul context key for a collection of ignored headers for the current request.
	 * Pre-filters can set this up as a set of lowercase strings.
	 */
	public static final String IGNORED_HEADERS = "ignoredHeaders";

This allowed "pre" filters to add ignoredHeaders to the context. You've swapped that out for a single list in `ProxyRequestHelper`.

I think it makes sense to have both, but you can't remove the old way of working.

**Comment by liebowitz** [comment]
Got it now, but it might be confusing to have both "globally" ignored headers and per-request ignored headers - I'll revert the change. 

If we added a getter for ProxyRequestHelper in SimpleHostRoutingFilter, it could be set in the request:
```
class CustomHostRoutingFilter extends SimpleHostRoutingFilter {
	@Override
	public Object run() {
		getHelper().addIgnoredHeaders("X-Ignored");
		return super.run();
	}
}
```
Or, delegate it to a protected method:
```
class CustomHostRoutingFilter extends SimpleHostRoutingFilter {
	@Override
	public Object run() {
		super.addIgnoredHeaders("X-Ignored");
		return super.run();
	}
}
```
Thoughts?

**Comment by spencergibb** [comment]
I'd prefer the delegate to a protected method.

**Comment by liebowitz** [comment]
Checked in.  Thanks.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/SimpleHostRoutingFilter.java [hunk 1]
```diff
@@ -138,9 +115,25 @@ public SimpleHostRoutingFilter(ProxyRequestHelper helper) {
...
+				}
+				connectionManager.closeExpiredConnections();
+			}
+		}, 30000, 5000);
+	}
+
 	@PreDestroy
 	public void stop() {
-		CONNECTION_MANAGER_TIMER.cancel();
+		CONNECTION_MANAGER_TIMER.purge();
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/SimpleHostRoutingFilter.java]: Why change this to `purge`?

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/SimpleHostRoutingFilter.java [hunk 2]
```diff
@@ -185,9 +177,66 @@ public Object run() {
 		return null;
 	}
 
+	protected PoolingHttpClientConnectionManager newConnectionManager() {
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/SimpleHostRoutingFilter.java]: Why did you move this as opposed to just removing static? I can't tell if you've made changes or just moved it.

Comment (liebowitz) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/SimpleHostRoutingFilter.java]: I moved it up so the diff would be clearer.

Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/SimpleHostRoutingFilter.java]: It's not clearer, IMO it's worse as I can't tell what changes (if any) you made to this method.

Comment (liebowitz) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/SimpleHostRoutingFilter.java]: The entire method body changed, now using the updated HttpClient builder API.

Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/SimpleHostRoutingFilter.java]: right, and you have to explain the changes, I can't see them side by side.

Comment (liebowitz) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/zuul/filters/route/SimpleHostRoutingFilter.java]: newConnectionManager()
- Switched from deprecated ThreadSafeClientConnManager to PoolingHttpClientConnectionManager (thread-safe)
- Switched from deprecated SchemeRegistry to RegistryBuilder
- Switched from deprecated SSLSocketFactory to SSLConnectionSocketFactory

newClient()
- Switched from deprecated DefaultHttpClient to CloseableHttpClient 
- Switched from deprecated HttpParams to RequestConfig class.

**Comment by spencergibb** [comment]
Merged with a little polish via c93a19c0d1fbfd34bbc1aba2b1b6223872f62589

**Pull request closed without merging**